torch/fx/experimental/symbolic_shapes.py

```diff
@@ -2302,15 +2304,18 @@ def _create_symbolic_sizes_strides_storage_offset(
                 for i in range(len(size))
                 if stride[i] is not None and ex_stride[i] >= 0
             }
+
             # iterate over unbound strides in sorted order
-            val_list = sorted(
-                [(ex_stride[i], i) for i in range(len(stride)) if stride[i] is None],
-                key=lambda tup: (
-                    # Order nested int by their coefficients.
-                    # 1 here to order nested int after non-nested int.
+            def _nested_int_aware_sort(tup):
+                return (
+                    # Order nested ints by their coefficients.
+                    # 1 here to order nested ints after non-nested-ints.
```

Comment:
One thing we do support today is temporary transposing of the ragged dimension for the purposes of passing to SDPA:
```python
# shape (B, j0, D)
nt = nested_tensor(...)
# sdpa requires shape with all batch dims first: (B_0, B_1, ..., seq_len, D)
# (B, j0, D) -> (B, j0, num_heads, head_dim) -> (B, num_heads, j0, head_dim)
nt = nt.unflatten(-1, [num_heads, head_dim]).transpose(-2, -3)
output = sdpa(nt, nt, nt)
```

This is modeled via `nt._ragged_idx` and supported for very few operations. I think this does necessitate full sorting here, but I may be wrong.

Reply:
@jbschlosser @ezyang FYI pushed some new work on fold `transpose()` that should shed some light, see [this notebook](https://github.com/bhosmer/fold/blob/main/notebooks/transpose.ipynb) for examples including the SDPA case. Full disclosure, I don't know what sorting means in this context, but re Ed's examples above:

* it's true that the first example `(3, n0, 5)` won't work if `n0` is a (non-degenerate) ragged dim. Dimensions aren't completely rigid under transposition, but the immediate compatibility requirement that gets violated in this case is that `n0`'s cycle (the length of its largest nonrepeating subsequence) must be <= the GCD of the dimensions being transposed (in this case 1, which would make it a rectangular dim).

* The second example is fine if (as Ed points out) you transpose the sequence(s) inward of the transposition, which fold does - well, the fold *model* does, the implementation hasn't been generalized to handle this case yet. E.g. `(2, 4, [1, 2, 3, 4, 1, 2, 3, 4]).transpose(0, 1)` will have shape `(4, 2, [1, 1, 2, 2, 3, 3, 4, 4])`. (Aside: abstractly this always happens, but if the dimensions inward of the transposition are rectangular it's invisible.) 

But to restate the obvious though, neither of these is the SDPA case: that's an adjacent ragged outer dim and rectangular inner dim getting swapped, and all dimensions inward of the swap are rectangular. (Example is near the top of the notebook.) Happy to dig into/talk about this more if it's useful, just LMK.

Reply:
@ezyang there's some discretion in which representations get picked when new dims are constructed, so this is some library function noticing that an Affine would work (sorted sequence with regular step size) and using it instead of a straight Seq. (And not checking sequence size to make sure it pays its way.)

Should be simple to add a toggle to dumb down this behavior so the outputs are less inscrutable, if that's useful (the place where the smart choices matter is shape transforms on big arrays, where you don't want the metadata size to explode, but for didactic purposes it's probably not useful).

You can also demystify it by running it through this
```
def expand(shape):
    def expand_dim(d):
        return d.w if isinstance(d, Rect) else [w for w in d]
    return tuple(expand_dim(d) for d in shape)
```
But yeah I'll also add some more examples to that notebook.

Reply:
@ezyang @jbschlosser FYI updated [notebook](https://github.com/bhosmer/fold/blob/main/notebooks/transpose.ipynb) with a bunch more examples. Also improved the implementation, e.g. Ed's second example above works now (it's in the notebook)